doc/rfcs/003-quickly-collaborate-with-coworkers-and-friends.md

```diff
@@ -0,0 +1,146 @@
+3. Host switches from Atom to third-party communication tool (e.g., Slack, IRC)
+4. Host pastes portal URL into third-party communication tool
+5. Guest clicks portal URL, and the operating system hands control to Atom, and Atom joins the portal
+
+With the ability to invite past collaborators to your portal from within Atom, we reduce the host's set-up process from 4 steps to 3 steps:
+
+1. Host shares a portal
+2. Host selects person in their list of past collaborators
+3. Host invites person to join the portal
+4. Guest accepts invitation to join portal, and Atom joins the portal
+
+And perhaps we can reduce the host's set-up process even further:
+
+1. Host selects a person in their list of past collaborators
+2. Host invites the selected person to a portal, and Atom automatically shares a portal for the host and invites the selected person
+3. Guest accepts invitation to join portal, and Atom joins the portal
+
+## Explanation
+
+### Share a portal via URL
+
+Once a host creates a portal (e.g., by clicking the "share" toggle in Teletype's status bar popover), Teletype presents the host with the portal's URL. The host copies that URL to their clipboard (in the same way that they previously copied the portal ID).
+
+#### Early mockup
+
+<img width="314" src="https://user-images.githubusercontent.com/2988/37489877-6d0b0cfe-286f-11e8-87a3-a04da42994a3.png">
+
+Also consider [Slack's approach](https://github.com/atom/teletype/issues/109#issuecomment-337223084) for potential inspiration.
+
+#### Portal URL format
+
+Prior Teletype releases established a URI structure for identifying editors in Atom's workspace: `atom://teletype/portal/<portal-id>/editor/<editor-id>`
+
+Portal URLs use that same pattern, but exclude the editor ID: `atom://teletype/portal/<portal-id>`
+
+The `<portal-id>` is the same ID used in prior releases for sharing and joining a portal.
+
+Example: `atom://teletype/portal/63b120f3-b646-4c46-8962-656518249186`
+
+### Join a portal via URL
+
+When a guest follows the URL (e.g., by clicking on the URL in Slack, IRC, etc.), Atom opens, and Teletype asks the user if they want to join the portal. If the user chooses to join the portal, they become a guest in the portal (just as they previously did by entering the portal ID and clicking "Join"). If they choose not to join the portal, nothing happens.
+
+To honor the [UX guidelines for Atom URI handlers](https://flight-manual.atom.io/hacking-atom/sections/handling-uris/), Teletype avoids automatically joining the portal. When asking the user whether they want to join the portal, Teletype offers an option to automatically join future portals. This option is disabled by default. When the user enables this option, any time they follow a portal URL, Teletype will automatically join the portal without the user having to perform an additional confirmation of their desire to join the portal. Users can disable this option at any time via the Teletype settings in Atom's Settings UI.
```

Comment:
When a guest joins a new portal by clicking on the "Join portal" button in the Atom notification, does Teletype open a new window or use the same window?

Now that multiple files are supported, I might not want them to get mixed with all the open tabs and fuzzy finder of my current workspace and rather have an empty window during this collaboration.

The notification could also have two options (buttons) to join a portal?

1. Join portal in this window
2. Join portal in a new window

Probably better would be a ✅ "Open in new window" checkbox that can be toggled, but I don't think the Atom notifications API supports having checkboxes?

Reply:
Thanks for bringing this up. :zap:

> The notification could also have two options (buttons) to join a portal?

I can see the benefit of that. If we can find a way to present the options nicely in the UI, then I'm all for it. We could offer 3 buttons, but that feels overcomplicated to me:

<img width="472" src="https://user-images.githubusercontent.com/2988/37614679-bb84194c-2b81-11e8-9029-3911ded8bcf9.png">

Suddenly I'm having flashbacks to every restaurant in California.

![easy as 1-2-3](https://user-images.githubusercontent.com/2988/37615024-aa3a79e6-2b82-11e8-957a-66f51fa5e706.jpg)

"So many decisions. Was this avocado toast really worth all this effort?" :innocent::laughing:

All joking aside, I'm tempted to ship the first release with just two buttons: "Join" and "Decline". Then, once we've used it for a while, and other folks have had a chance to use it as well, we can assess the value/priority of offering other options. What do you think?

Reply:
Opening a new window is going to be non-trivial to implement, and the window will also open more slowly than would be ideal due to Atom's load time issues. That said, I do think that a remote workspace really *should* appear in its own window. I think the new window option should be considered out of scope for this RFC.

Reply:
> Suddenly I'm having flashbacks to every restaurant in California.

:joy:

Not sure if technically possible, but this might work:

1. You get a notification to join a portal.
2. You press `cmd-shift-n` to open a new window.
3. The notification appears in the new window too and you click "join portal".

Reply:
> When the user enables this option, any time they follow a portal URL, Teletype will automatically join the portal without the user having to perform an additional confirmation of their desire to join the portal.

Is this going to be specific per user or is this going to be a global configuration? My preference would be per user (i.e. "Would you like to automatically join any sessions with `@UserA`?") as opposed to a global configuration (i.e. "Automatically join all Teletype pairing sessions for all time")